x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -381,14 +381,24 @@ public void deleteTrainedModel(String modelId, ActionListener<Boolean> listener)
     public void expandIds(String idExpression,
                           boolean allowNoResources,
                           @Nullable PageParams pageParams,
+                          Set<String> tags,
                           ActionListener<Tuple<Long, Set<String>>> idsListener) {
         String[] tokens = Strings.tokenizeToStringArray(idExpression, ",");
+        BoolQueryBuilder tagQuery = QueryBuilders.boolQuery();
+        for(String tag : tags) {
+            tagQuery.filter(QueryBuilders.termQuery(TrainedModelConfig.TAGS.getPreferredName(), tag));
+        }
+
+        QueryBuilder query = QueryBuilders.constantScoreQuery(
+            QueryBuilders.boolQuery()
+                .filter(tagQuery.hasClauses() ? tagQuery : QueryBuilders.matchAllQuery())
```

Comment:
Alternatively you could drop this `filter` part altogether if there are no tags.
One more thought: would the following query be equivalent to what you have right now:
```
QueryBuilder query = QueryBuilders.boolQuery()
    .filter(buildQueryIdExpressionQuery(tokens, TrainedModelConfig.MODEL_ID.getPreferredName())));
for (String tag : tags) {
  query.filter(QueryBuilders.termQuery(TrainedModelConfig.TAGS.getPreferredName(), tag));
}
```
?
This way we can get rid of the ternary operator `?` altogether.

Reply:
Ah, I do think your query building is better. Will see about refactoring :)